pkg/provisioner/provision.go

```diff
@@ -73,28 +73,44 @@ func (p *ZFSProvisioner) Provision(ctx context.Context, options controller.Provi
 		},
 		Spec: v1.PersistentVolumeSpec{
 			PersistentVolumeReclaimPolicy: reclaimPolicy,
-			AccessModes:                   []v1.PersistentVolumeAccessMode{v1.ReadWriteMany, v1.ReadOnlyMany, v1.ReadWriteOnce},
+			AccessModes:                   createAccessModes(options, useHostPath),
 			Capacity: v1.ResourceList{
 				v1.ResourceStorage: options.PVC.Spec.Resources.Requests[v1.ResourceStorage],
 			},
-			PersistentVolumeSource: createVolumeSource(parameters, dataset),
-			NodeAffinity:           createNodeAffinity(parameters),
+			PersistentVolumeSource: createVolumeSource(parameters, dataset, useHostPath),
+			NodeAffinity:           createNodeAffinity(parameters, useHostPath),
 		},
 	}
 	return pv, controller.ProvisioningFinished, nil
 }
 
-func createVolumeSource(parameters *ZFSStorageClassParameters, dataset *zfs.Dataset) v1.PersistentVolumeSource {
-	if parameters.NFS != nil {
-		return v1.PersistentVolumeSource{
-			NFS: &v1.NFSVolumeSource{
-				Server:   parameters.Hostname,
-				Path:     dataset.Mountpoint,
-				ReadOnly: false,
-			},
+func canUseHostPath(parameters *ZFSStorageClassParameters, options controller.ProvisionOptions) bool {
+	switch parameters.Type {
+	case Nfs:
+		return false
+	case HostPath:
+		return true
+	case Auto:
+		if slices.Contains(options.PVC.Spec.AccessModes, v1.ReadOnlyMany) || slices.Contains(options.PVC.Spec.AccessModes, v1.ReadWriteMany) {
+			return false
 		}
 	}
-	if parameters.HostPath != nil {
+	return true
+}
```

Comment:
There is one subtlety to take into account though: As far as I could understand it is perfectly possible to specify both ReadWriteOnce and ReadWriteMany in a volume claim. In that case, the provisionner is left to decide which of HostPath or NFS it prefers.
By inverting this condition, HostPath effectively becomes the default, whereas NFS was the default before.

Reply:
Unless you really meant:

> Only return true if !ROX && !RWX

instead of:

> Only return true if RWO/RWOP, otherwise false.

Because that would make more sense to leave NFS the default.

Reply:
You're right :grinning: 
It should be correct now with your latest change